### PR TITLE
[FIX] sale_management: options_lines have correct discount

### PR DESCRIPTION
**Pull request by fdamhaut**
Step to reproduce:
- Create a pricelist with 20% discount and
 Show discount to the user
- In a SO with public pricelist, add option line OL
- Change pricelist to the one created earlier

Current behaviour:
- OL unit price is 80% of list price
- OL discount is 0%

Behaviour after PR:
- OL unit price is 100% of list price
- OL discount is 20%

opw-2761252

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

**Review by Feyensv — COMMENTED** [review]
A little test please? :pray:

**Comment by fdamhaut** [comment]
@Feyensv 
TEST : :heavy_check_mark:

File: addons/sale_management/models/sale_order.py [hunk 1]
```diff
@@ -87,7 +87,7 @@ def update_prices(self):
         self.ensure_one()
         res = super().update_prices()
         for line in self.sale_order_option_ids:
-            line.price_unit = self.pricelist_id.get_product_price(line.product_id, line.quantity, self.partner_id, uom_id=line.uom_id.id)
+            line.write(self._compute_option_data_for_template_change(line))
```
Comment (Feyensv) [inline addons/sale_management/models/sale_order.py:90]: After a deeper analysis, `_compute_option_data_for_template_change` was made to receive `sale.order.template.option` records, not `sale.order.option` ones, even if this miraculously works, you should avoid using this method in this case (also, that way, the dict you are giving to write contains unnecessary information which shouldn't be updated, even if it's the same values).

I'd prefer if you follow the logic of `_onchange_product_id` below (you could also make a separate method called by both the onchange and `update_prices`).
```suggestion
        # To compute the discount a so line is created in cache
        values = line._get_values_to_add_to_order()
        new_sol = line.env['sale.order.line'].new(values)
        new_sol._onchange_discount()
        line.discount = new_sol.discount
        if line.order_id.pricelist_id and line.order_id.partner_id:
            line.price_unit = new_sol._get_display_price(product)
```

File: addons/sale_management/models/sale_order.py [hunk 2]
```diff
@@ -208,6 +207,24 @@ class SaleOrderOption(models.Model):
...
+            if pricelist:
+                pricelist_price = pricelist.with_context(uom=option.uom_id.id).get_product_price(
+                    option.product_id, 1, False)
+
+                if pricelist.discount_policy == 'without_discount' and price:
+                    discount = max(0, (price - pricelist_price) * 100 / price)
+                else:
+                    price = pricelist_price
+
+            option.write({'price_unit': price, 'discount': discount})
```
Comment (Feyensv) [inline addons/sale_management/models/sale_order.py:226]:
```suggestion
        values = option._get_values_to_add_to_order()
        new_sol = option.env['sale.order.line'].new(values)
        new_sol._onchange_discount()
        option.discount = new_sol.discount
        if option.order_id.pricelist_id and option.order_id.partner_id:
            option.price_unit = new_sol._get_display_price(product)
```

Why didn't you copy the logic below ?

Comment (fdamhaut) [reply addons/sale_management/models/sale_order.py]: I looked to see if there was other place which my new function could replace but glossed over this as there was not the `without_discount` line i searched for :facepalm: 
Should be better now :D

File: addons/sale_management/models/sale_order.py [hunk 3]
```diff
@@ -208,6 +207,26 @@ class SaleOrderOption(models.Model):
...
+                uom=option.uom_id.id,
+                fiscal_position=option.env.context.get('fiscal_position')
+            )
+            # To compute the discount a so line is created in cache
+            values = option._get_values_to_add_to_order()
+            new_sol = option.env['sale.order.line'].new(values)
+            new_sol._onchange_discount()
+            option.discount = new_sol.discount
+            if option.order_id.pricelist_id and option.order_id.partner_id:
+                option.price_unit = new_sol._get_display_price(product)
```
Comment (Feyensv) [inline addons/sale_management/models/sale_order.py:228]:
```suggestion
            # To compute the discount a so line is created in cache
            values = option._get_values_to_add_to_order()
            new_sol = option.env['sale.order.line'].new(values)
            new_sol._onchange_discount()
            option.discount = new_sol.discount
            if option.order_id.pricelist_id and option.order_id.partner_id:
                product = option.product_id.with_context(
                    partner=option.order_id.partner_id,
                    quantity=option.quantity,
                    date=option.order_id.date_order,
                    pricelist=option.order_id.pricelist_id.id,
                    uom=option.uom_id.id,
                    fiscal_position=option.env.context.get('fiscal_position')
                )
                option.price_unit = new_sol._get_display_price(product)
```
Well, now that we move the code, let's be strict and define the variable only in its scope of usage :)

File: addons/sale_management/tests/test_sale_order.py [hunk 4]
```diff
@@ -283,3 +283,57 @@ def test_03_template_with_discount_excluded_pricelist(self):
...
+        self.assertEqual(
+            self.sale_order.sale_order_option_ids[0].price_unit,
+            self.pub_option_price,
+            "If no pricelist is set, the unit price shall be the public option price.")
+
+        self.assertEqual(
+            self.sale_order.sale_order_option_ids[0].discount, 0,
+            "If no pricelist is set, the discount should be 0.")
+
+
```
Comment (Feyensv) [inline addons/sale_management/tests/test_sale_order.py:307]: useless double line jump ;)

File: addons/sale_management/tests/test_sale_order.py [hunk 5]
```diff
@@ -283,3 +283,57 @@ def test_03_template_with_discount_excluded_pricelist(self):
             "If a pricelist is set without discount included, the discount "
             "shall be computed according to the price unit and the subtotal."
             "price")
+
+    def test_04_update_pricelist_option_line(self):
+        """
+        This test checks that option line's data is correctly
+        updated when we update the pricelist
```
Comment (Feyensv) [inline addons/sale_management/tests/test_sale_order.py:290]:
```suggestion
        updated when we the prices are recomputed after a pricelist update
```

File: addons/sale_management/tests/test_sale_order.py [hunk 6]
```diff
@@ -283,3 +283,57 @@ def test_03_template_with_discount_excluded_pricelist(self):
             "If a pricelist is set without discount included, the discount "
             "shall be computed according to the price unit and the subtotal."
             "price")
+
+    def test_04_update_pricelist_option_line(self):
+        """
+        This test checks that option line's data is correctly
```
Comment (Feyensv) [inline addons/sale_management/tests/test_sale_order.py:289]:
```suggestion
        This test checks that option line's values are correctly
```

File: addons/sale_management/models/sale_order.py [hunk 7]
```diff
@@ -208,6 +207,26 @@ class SaleOrderOption(models.Model):
     quantity = fields.Float('Quantity', required=True, digits='Product Unit of Measure', default=1)
     sequence = fields.Integer('Sequence', help="Gives the sequence order when displaying a list of optional products.")
 
+    def _update_price_discount(self):
```
Comment (Feyensv) [inline addons/sale_management/models/sale_order.py:210]:
```suggestion
    def _update_price_and_discount(self):
```

would be more clear imho :)

File: addons/sale_management/tests/test_sale_order.py [hunk 8]
```diff
@@ -283,3 +283,57 @@ def test_03_template_with_discount_excluded_pricelist(self):
...
+        self.sale_order.write({
+            'pricelist_id': self.discount_included_price_list.id,
+        })
+        self.sale_order.update_prices()
+
+        self.assertEqual(
+            self.sale_order.sale_order_option_ids[0].price_unit,
+            self.pl_option_price,
+            "If a pricelist is set with discount included,"
+            " the unit price shall be the public product price.")
```
Comment (Feyensv) [inline addons/sale_management/tests/test_sale_order.py:317]:
```suggestion
            " the unit price shall be the discounted price.")
```

File: addons/sale_management/tests/test_sale_order.py [hunk 9]
```diff
@@ -283,3 +283,57 @@ def test_03_template_with_discount_excluded_pricelist(self):
...
+
+        self.sale_order.write({
+            'sale_order_template_id': self.quotation_template_no_discount.id
+        })
+        self.sale_order.onchange_sale_order_template_id()
+
+        self.assertEqual(
+            self.sale_order.sale_order_option_ids[0].price_unit,
+            self.pub_option_price,
+            "If no pricelist is set, the unit price shall be the public option price.")
```
Comment (Feyensv) [inline addons/sale_management/tests/test_sale_order.py:301]:
```suggestion
            "If no pricelist is set, the unit price shall be the product sales price.")
```

File: addons/sale_management/tests/test_sale_order.py [hunk 10]
```diff
@@ -283,3 +283,57 @@ def test_03_template_with_discount_excluded_pricelist(self):
...
+        self.sale_order.write({
+            'pricelist_id': self.discount_excluded_price_list.id,
+        })
+        self.sale_order.update_prices()
+
+        self.assertEqual(
+            self.sale_order.sale_order_option_ids[0].price_unit,
+            self.pub_option_price,
+            "If a pricelist is set without discount included,"
+            " the unit price shall be the public product price.")
```
Comment (Feyensv) [inline addons/sale_management/tests/test_sale_order.py:333]:
```suggestion
            " the unit price shall be the product sales price.")
```

**Review by Feyensv — CHANGES_REQUESTED** [review]
nearly good to go :+1:

**Review by Feyensv — APPROVED** [review]
@robodoo r+

File: addons/sale_management/models/sale_order.py [hunk 11]
```diff
@@ -226,22 +245,9 @@ def _onchange_product_id(self):
...
             lang=self.order_id.partner_id.lang,
-            partner=self.order_id.partner_id,
-            quantity=self.quantity,
-            date=self.order_id.date_order,
-            pricelist=self.order_id.pricelist_id.id,
-            uom=self.uom_id.id,
-            fiscal_position=self.env.context.get('fiscal_position')
         )
         self.name = product.get_product_multiline_description_sale()
-        self.uom_id = self.uom_id or product.uom_id
```
Comment (Feyensv) [inline addons/sale_management/models/sale_order.py:237]: Oopsy, we lost the uom change :man_facepalming:

**Pull request closed without merging**